apollo-router/src/services/router_service.rs

```diff
@@ -0,0 +1,1157 @@
+                                    // Apparently HTTP/2 trailers are like headers, except after the response body.
+                                    // I (Simon) believe nothing in the Apollo Router uses trailers as of this writing,
+                                    // so ignoring `poll_trailers` is fine.
+                                    // If we want to use trailers, we may need remove this convertion to `hyper::Body`
+                                    // and return `UnsyncBoxBody` (a.k.a. `axum::BoxBody`) as-is.
+                                });
+
+                            Ok(RouterResponse { response, context })
+                        } else {
+                            // this should be unreachable due to a previous check, but just to be sure...
+                            Ok(router::Response {
+                                response: http::Response::builder()
+                                    .status(StatusCode::NOT_ACCEPTABLE)
+                                    .body(Body::from(
+                                        format!(
+                                            r#"'accept' header can't be different from \"*/*\", {:?}, {:?} or {:?}"#,
+                                            APPLICATION_JSON_HEADER_VALUE,
+                                            GRAPHQL_JSON_RESPONSE_HEADER_VALUE,
+                                            MULTIPART_DEFER_CONTENT_TYPE
+                                        )
+                                    ))
+                                    .expect("cannot fail"),
+                                context,
+                            })
+                        }
+                    }
+                }
+            } else {
+                // BAD REQUEST
+                Ok(router::Response {
+                    response: http::Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(Body::from("Invalid GraphQL request"))
+                        .expect("cannot fail"),
+                    context,
+                })
+            }
+        };
+        Box::pin(fut)
+    }
+}
+
+// Process the headers to make sure that `VARY` is set correctly
+fn process_vary_header(headers: &mut HeaderMap<HeaderValue>) {
+    if headers.get(VARY).is_none() {
+        // We don't have a VARY header, add one with value "origin"
+        headers.insert(VARY, HeaderValue::from_static("origin"));
+    }
+}
+
+/// A collection of services and data which may be used to create a "router".
+#[derive(Clone)]
+pub(crate) struct RouterCreator<SF>
+where
+    SF: ServiceFactory<supergraph::Request> + Clone + Send + Sync + 'static,
+{
+    supergraph_creator: Arc<SF>,
+    static_page: StaticPageLayer,
+}
+
+impl<SF> ServiceFactory<router::Request> for RouterCreator<SF>
+where
+    SF: StuffThatHasPlugins + ServiceFactory<supergraph::Request> + Clone + Send + Sync + 'static,
```

Comment:
Am I the only one to find this name `StuffThatHasPlugins` a little bit weird ? :p

Reply:
I suggest to rename it to `HasPlugins`

Reply:
Haha fair, FWIW I forgot to rename it, let's go with HasPlugins